File: TP1/goodpresident.py
def DFS_MOD(n_cities, c_libs, c_roads, graph):
    #n_cities = min(n_cities, 100000)
    lib_counter = 0
    visited = {u: False for u in graph.keys()}  
    for city in graph.keys():
        if not visited[city]:
            lib_counter += 1
            visited = DFS_visited(graph, city, visited)
            
    return lib_counter * c_libs + (n_cities-lib_counter) * c_roads

def DFS_visited(graph, u, visited):
    visited[u] = True
    for v in graph[u]:
        if not visited[v]:
            DFS_visited(graph, v, visited)
    return visited

def run_president(n_cities, n_roads, c_libs, c_roads):
    if n_roads == 0:
        return c_libs * n_cities

    roads = []
    
    for _ in range(n_roads):
        roads.append(tuple(map(int, input().split())))
        
        
    if c_libs <= c_roads:
        return c_libs * n_cities
    
    graph = {city: set() for city in range(1, n_cities + 1)}
    for road in roads:
        s,t = road[0], road[1]
        graph[s].add(t)
        graph[t].add(s)


    return DFS_MOD(n_cities, c_libs, c_roads, graph)

T = int(input())
for _ in range(min(T, 10)):
    n_cities, n_roads, c_libs, c_roads = map(int, input().split())
    print(run_president(n_cities, n_roads, c_libs, c_roads))

File: TP1/goodpresident.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <stack>

using namespace std;

long long DFS_MOD(int n_cities, int c_libs, int c_roads, vector<unordered_set<int>>& graph) {
    long long lib_counter = 0;
    unordered_set<int> visited;
    for (int city = 1; city <= n_cities; ++city) {
        if (visited.find(city) == visited.end()) {
            ++lib_counter;
            stack<int> stack;
            stack.push(city);
            while (!stack.empty()) {
                int current_city = stack.top();
                stack.pop();
                if (visited.find(current_city) == visited.end()) {
                    visited.insert(current_city);
                    for (int neighbor : graph[current_city]) {
                        if (visited.find(neighbor) == visited.end()) {
                            stack.push(neighbor);
                        }
                    }
                }
            }
        }
    }
    return lib_counter * c_libs + (visited.size() - 1) * c_roads;
}

long long run_president(int n_cities, int n_roads, int c_libs, int c_roads) {
    if (n_roads == 0) {
        return (long long) c_libs * n_cities;
    }

    vector<unordered_set<int>> graph(n_cities + 1);
    for (int i = 0; i < n_roads; ++i) {
        int s, t;
        cin >> s >> t;
        graph[s].insert(t);
        graph[t].insert(s);
    }
    
    if (c_libs <= c_roads) {
        return (long long) c_libs * n_cities;
    }

    return DFS_MOD(n_cities, c_libs, c_roads, graph);
}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T && i < 10; ++i) {
        int n_cities, n_roads, c_libs, c_roads;
        cin >> n_cities >> n_roads >> c_libs >> c_roads;
        cout << run_president(n_cities, n_roads, c_libs, c_roads) << endl;
    }
    return 0;
}
